assemblers/tcp_stream.go

```diff
@@ -85,22 +103,71 @@ func (stream *tcpStream) Accept(tcp *layers.TCP, ci gopacket.CaptureInfo, dir re
 	return accept
 }
 
+// Reassembled is an implementation of the reassembly.Stream interface
 func (stream *tcpStream) ReassembledSG(sg reassembly.ScatterGather, ac reassembly.AssemblerContext) {
+	// Get the direction of the packet (client to server or server to client)
 	dir, _, _, _ := sg.Info()
-	if dir == reassembly.TCPDirClientToServer {
-		stream.client.reassembledSG(sg, ac)
+	isClient := dir == reassembly.TCPDirClientToServer
+
+	// get our custom context that includes the TCP seq/ack numbers
+	ctx, ok := ac.(*Context)
+	if !ok {
+		log.Warn().
+			Msg("Failed to cast ScatterGather to ContextWithSeq")
+	}
+
+	// We use TCP SEQ & ACK numbers to identify request/response pairs
+	// A request's ACK corresponds to the SEQ of it's matching response
+	// https://madpackets.com/2018/04/25/tcp-sequence-and-acknowledgement-numbers-explained/
+	var requestId int64
+	if isClient {
+		requestId = int64(ctx.ack)
 	} else {
-		stream.server.reassembledSG(sg, ac)
+		requestId = int64(ctx.seq)
+	}
+
+	// get the number of packets that made up this request
+	packetCount := sg.Stats().Packets
+
+	// get the data from the packet
+	len, _ := sg.Lengths()
+	data := sg.Fetch(len)
+
+	// reset the buffer reader to use the new packet data
+	// bufio.NewReader creates a new 16 byte buffer on each call which we want to avoid
+	// https://github.com/golang/go/blob/master/src/bufio/bufio.go#L57
+	stream.buffer.Reset(bytes.NewReader(data))
+
+	// loop through the parsers until we find one that can parse the request/response
+	for _, parser := range stream.parsers {
+		success, err := parser.parse(stream, requestId, ctx.CaptureInfo.Timestamp, isClient, stream.buffer, packetCount)
+		if err != nil {
+			// if we hit the end of the stream, stop trying to parse
+			if err == io.EOF || err == io.ErrUnexpectedEOF {
+				break
+			}
+			log.Debug().
+				Err(err).
+				Int64("request_id", requestId).
+				Str("stream_ident", stream.ident).
+				Str("src_ip", stream.srcIP).
+				Str("src_port", stream.srcPort).
+				Str("dst_ip", stream.dstIP).
+				Str("dst_port", stream.dstPort).
+				Msg("Error parsing packet")
+			continue
+		}
+		if success {
+			break
+		}
 	}
 }
 
-// ReassemblyComplete is called when the TCP assembler believes a stream has completed.
+// ReassemblyComplete is an implementation of the reassembly.Stream interface
```

Comment:
```suggestion
// ReassemblyComplete implements gopacket's [reassembly.Stream.ReassemblyComplete] interface.
// Called when gopacket's assembly internals has decided that there is no more data for this Stream
// (e.g. FIN or RST packet, timed out without new data).
//
// Our implementation always returns true to remove the connection from the gopacket-managed pool.
// We don't return false, because we aren't interested in the last FIN-ACK.
```